testing/drivertest/drivertest_mps2_zerointerrupt.c

```diff
@@ -0,0 +1,270 @@
+
+#include <nuttx/nuttx.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <setjmp.h>
+#include <string.h>
+#include <cmocka.h>
+
+#include <nuttx/arch.h>
+#include <nuttx/irq.h>
+#include <nuttx/semaphore.h>
+
+#include <pthread.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define CONFIG_TEST_IRQPRIO_TTHREAD  8
+#define CONFIG_TEST_IRQPRIO_LOOP_CNT 5000
+
+#ifndef NVIC_IRQ_PENDSV
+#  define NVIC_IRQ_PENDSV 14
+#endif
+
+#define MPS2_ADDR2REG_PTR(base, off) (uint32_t*)((uint32_t*)(base) + (off))
+#define MPS2_IRQ_FROMBASE(base, off) ((base) + (off))
+
+/* https://developer.arm.com/documentation/101104/0200/programmers-model/
+ * base-element/cmsdk-timer
+ */
+
+#define MPS_TIMER_CTRL_OFFSET   0
+#define MPS_TIMER_VALUE_OFFSET  1
+#define MPS_TIMER_RELOAD_OFFSET 2
+#define MPS_TIMER_CLEAR_OFFSET  3
+
+#define MPS_TIMER_CTRL_ENABLE   (1<<0)
+#define MPS_TIMER_CTRL_IE       (1<<3)
+
+static_assert(NVIC_SYSH_PRIORITY_DEFAULT == 0x80, "prio");
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+typedef struct mps2_an500_timer_s
+{
+  volatile uint32_t *reload;
+  volatile uint32_t *ctrl;
+  volatile uint32_t *clear;
+  int irq;
+  int before;
+  sem_t *sem;
+  int after;
+} mps2_an500_timer_t;
+
+/****************************************************************************
+ * Private Functions Prototypes
+ ****************************************************************************/
+
+static int pendsv_irq_handle(int irq, void *context, void *arg);
+static int timer_irq_handle(int irq, void *context, void *arg);
+static int g_tag = 0;
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static mps2_an500_timer_t g_timer;
+static sem_t test_sem = SEM_INITIALIZER(0);
+
+static const int armv7m_gpio_base = 16;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static void mps_timer_init(void)
+{
+  static const uint32_t timerbase[] =
+    {
+      0x40000000, 0x40001000
+    };
+
+  static const int timerirq[] =
+    {
+      8, 9
+    };
+
+  static const int timer_irq_prio[] =
+    {
+      0x60, 0x70
+    };
+
+  mps2_an500_timer_t *t = &g_timer;
+
+  t->reload  = MPS2_ADDR2REG_PTR(timerbase[0], MPS_TIMER_RELOAD_OFFSET);
+  t->ctrl  = MPS2_ADDR2REG_PTR(timerbase[0], MPS_TIMER_CTRL_OFFSET);
+  t->clear = MPS2_ADDR2REG_PTR(timerbase[0], MPS_TIMER_CLEAR_OFFSET);
+  t->irq   = MPS2_IRQ_FROMBASE(armv7m_gpio_base, timerirq[0]);
```

Comment:
@GUIDINGLI I think it will work only in FLAT mode, right?

Reply:
Yes

Reply:
Maybe this information should be displayed to user, to avoid issues. Or the Kconfig needs to include depends on FLAT mode.

Reply:
MPS2-AN500 doesn't support the protected mode yet, so it isn't a problem now.